### PR TITLE
[trainer] fix: exclude the never-routed final response token from rollout MoE load-balance metrics

### PR DESCRIPTION
**Pull request by zsnoob**
### What does this PR do?

Fixes a systematic bias in the rollout MoE load-balance metrics introduced in #6853.

The last generated token of each sequence (EOS / max-length stop) is sampled but never fed back through the model, so it has **no routing record**: SGLang's capturer intentionally returns `seqlen - 1` records per request (see sgl-project/sglang#20964 for the design), and `AgentLoopWorker._postprocess` zero-fills the assembly buffer, leaving zeros at each sequence's last valid response position.

`_compute_rollout_moe_load_counts` selects response tokens via `response_mask`, which marks that position as valid (it is a real token — it just has no routing data), and bincounts the filler. Every sequence therefore credits expert 0 with `num_layers * topk` phantom assignments:

- expert 0's measured load is inflated by roughly `num_experts / avg_response_len` — ~6% relative at 2k response length with 128 experts, ~25% at 512, identically in every layer;
- the bias does not average out (every sequence contributes it) and looks exactly like a routing hotspot on expert 0 — in a metric whose purpose is to measure imbalance.

This PR drops each sequence's last valid response position before counting. It is the metrics-side counterpart of `build_r3_replay_mask` (#6473), which skips the same row on the replay side ("skip only the final response row, whose logits are not used"). It also makes the metric robust to a future negative sentinel filler, which today would silently disable the whole metric every step via the expert-id range guard.

### Checklist Before Starting

- [x] Search for similar PRs: https://github.com/verl-project/verl/pulls?q=is%3Apr+routed_experts+metrics
- [x] Format the PR title as `[{modules}] {type}: {description}`

### Test

`tests/trainer/ppo/test_rollout_moe_lb_metrics_on_cpu.py`:

- Existing fixtures updated to carry an explicit trailing filler slot; **all original assertions are unchanged**, demonstrating that the same real records plus the structural filler row produce the same metrics.
- New tests:
  - `..._last_token_filler_is_ignored`: counts are independent of the filler value (0 or junk); no phantom expert-0 load.
  - `..._survive_negative_filler`: a `-1` sentinel in the never-routed slot no longer disables the whole metric via the range guard.
  - `..._drop_last_valid_per_sequence`: ragged batches drop each sequence's own final valid position, not a shared column.
  - `..._single_token_response_yields_nothing`: a one-token response consists solely of the never-routed token and contributes no counts.

Verification: without the fix, 8 of the 14 tests fail; with it, 14/14 pass.

### API and Usage Example

No API change. Metric keys and shapes are unchanged; only the counted token set is corrected.

### Design & Code Changes

- `verl/trainer/ppo/metric_utils.py`: in `_compute_rollout_moe_load_counts`, compute each row's last valid index from `response_mask` (vectorized, handles ragged masks and rows with no valid tokens) and clear it on a cloned mask before selecting tokens.
- `tests/trainer/ppo/test_rollout_moe_lb_metrics_on_cpu.py`: fixture updates + 4 new tests as above.

### Checklist Before Submitting

- [x] Read the Contribute Guide.
- [x] Apply pre-commit checks (ruff check + format pass on the changed files).
- [ ] Add / Update the documentation — not applicable (internal metric correctness).
- [x] Covered by unit tests in the existing CPU CI test file.

### CODE REVIEW

File: verl/trainer/ppo/metric_utils.py [hunk 1]
```diff
@@ -230,6 +235,17 @@ def _compute_rollout_moe_load_counts(
...
+    # so it has no routing record; its slot holds filler from batch assembly
+    # (zeros, see AgentLoopWorker._postprocess). Counting it would credit
+    # expert 0 with num_layers * topk phantom assignments per sequence, so drop
+    # each sequence's last valid position. This is the metrics counterpart of
+    # build_r3_replay_mask, which skips the same row on the replay side.
+    positions = torch.arange(response_len, device=response_mask.device)
+    last_valid = torch.where(response_mask, positions, positions.new_full((), -1)).amax(dim=-1)
+    has_valid = last_valid >= 0
+    response_mask = response_mask.clone()
+    response_mask[has_valid.nonzero(as_tuple=True)[0], last_valid[has_valid]] = False
```
[the remark this answers is not in the document]
Comment (zsnoob) [reply verl/trainer/ppo/metric_utils.py]: - Added the `response_len == 0` early return next to the other input guards
  (it also covers the pre-existing `routed_experts[:, -0:]` slice hazard),
  with a regression test.
- Replaced the `nonzero()` index write with the broadcast comparison mask as
  suggested; the defensive `clone()` is gone too since `&` allocates a fresh
  tensor.

**Comment by zsnoob** [comment]
CI note: cpu_unit_tests (which covers this change) passes. The 7 failing jobs
are unrelated to this diff — 5 are Ascend/NPU infra jobs, and the two GPU e2e
failures are environment flakes: megatron-qwen3 finished training successfully
and failed in `ray stop` teardown (psutil pidfd_open OSError), and
fsdp-qwen2_5vl-3b (a dense VLM, no MoE path) died at SGLang server
initialization with an OOM-killer signature. Happy to rebase/rerun if needed.

**Review by Luosuu — APPROVED** [review]
